### PR TITLE
Manage clients per broker node ID

### LINKED ISSUE #47: KafkaUnavailableError
The afkak client and producer are used for opencord/voltha kafka access.   We are seeing every 10+ minutes the following error:Unable to load metadata from configured hosts: <twisted.python.failure.Failure afkak.common.KafkaUnavailableError: All servers ([<KafkaBrokerClient 192.168.0.202:9092 clientId=u'afkak-client' connected>]) failed to process request>

When the error occurs there is no indication in the kafka-docker logs of any problem/error.
When this occurs the voltha/kafka_proxy stop /restart is generated which results in:

[2018-10-03 17:49:11,552] INFO [GroupMetadataManager brokerId=1001] Removed 0 expired offsets in 1 milliseconds. (kafka.coordinator.group.GroupMetadataManager)

Which is an expected message.

This error occurs when using afkak 3.0.0.dev0 and 2.4.0.

### LINKED ISSUE #59: KafkaClient.close() should always return a deferred
`KafkaClient.close()` sometimes returns `None` right now (specifically, when there are no broker clients). It should always return a `Deferred`.

### PR DESCRIPTION
**Pull request by twm**
# Overview

This PR makes several fundamental changes to fix #47:

1. The `KafkaClient.clients` map of `_KafkaBrokerClient` objects ("broker clients") is now keyed by broker node ID rather than (host, port).

    `KafkaClients.clients` was keyed by broker (host, port) before. However, this map was populated based on two services. the (host, port) given in Kafka metadata responses doesn't necessarily match the `hosts` argument passed

2. Clients are created and added to this map lazily and broker clients connect to brokers lazily (only as required to handle requests). Connections are no longer 

    This reduces churn when cluster metadata changes.

    This improves efficiency because a client used for a single purpose may only need to interact with a single broker. For example, a producer always messages the leader for the partition(s) it produces to. If there is a single partition, only a single connection is required. If the cluster is large, that may be a small subset of the known brokers.

3. Connections to brokers are no longer closed when the (host, port) associated with that broker changes.

    Brokers may validly be accessible by more than one address. Consider the case of a rolling restart to install new hostnames as `advertised.listeners` (which [is a list](https://kafka.apache.org/documentation/#brokerconfigs)).

    Once the connection drops the broker client will reconnect to the last observed address.

3. A special "bootstrap" client is used to make metadata requests before the brokers in the cluster are known.

    The bootstrap client makes an ephemeral connection to a hostname from the `hosts` argument (now internally referred to as `_bootstrap_hosts`). It also serves as a fallback when making broker unaware requests (metadata requests, consumer group metadata requests, etc.).

There are also some other notable changes:

4. `KafkaClient` no longer clears metadata when a broker disconnects. This may be controversial, but I am concerned that this violates the expectations of methods like `send_produce_request()` that load metadata and then make requests based on it, creating a race condition.

5. `KafkaClient.close()` now more aggressively poisons the client against reuse. It also always returns a `Deferred` (closes #59).

6. Endpoints are used to connect to the broker. The default is `HostnameEndpoint`, but this may be overridden by passing the *endpoint_factory* argument to `KafkaClient`.

# Mocking I/O: endpoints and `Connections`

The endpoints change enables a change in unit testing strategy. The new `afkak.test.endpoints.Connections` class serves as an *endpoint_factory* that can accept/reject connections and even respond to requests. Its use looks like this:

```python
connections = Connections()
client = KafkaClient(hosts='kafka', endpoint_factory=connections)

d = client.load_metadata_for_topics(...)  # triggers a connection request

conn = client.accept('kafka')  # Match the hostname. Can also use a pattern like `'*'`.
conn.pump.flush()  # Move I/O from client to server.
request = self.successResultOf(conn.server.expectRequest(
    api_key=KafkaCodec.METADATA_KEY,
    api_version=0,
    correlation_id=1,  # or mock.ANY if you don't care
))
request.respond(b'...')  # The correlation_id is automatically prepended.
conn.pump.flush()  # Move I/O from server to client.

result = self.successResultOf(d)  # Now this has fired.
```

This replaces the technique of overriding `KafkaClient.clients` to contain `mock.Mock`. It is less vulnerable to mock typos and implicitly gets things like matching correlation IDs correct.

### CODE REVIEW

File: afkak/client.py [hunk 1]
```diff
@@ -584,23 +588,26 @@ def _handle_responses(self, responses, fail_on_error, callback=None,
                 out.append(resp)
         return out
 
-    def _get_brokerclient(self, host, port):
+    def _get_brokerclient(self, node_id):
         """
         Get or create a connection to a broker using host and port.
```
Comment (rthille) [inline afkak/client.py]: Comment need update to 'using broker node-id' or similar.

File: afkak/client.py [hunk 2]
```diff
@@ -652,42 +660,61 @@ def _clean_close_dlist(result, close_dlist):
...
+            to_close.append(self._clients.pop(node_id))
 
-        # Disconnect and remove from self.clients any removed brokerclients
+        # Remove any clients for brokers which no longer exist.
         if remove and removed_brokers:
-            self._close_brokerclients(removed_brokers)
+            for node_id in removed_brokers:
+                to_close.append(self.clients.pop(node_id))
+
+        # Identify any clients with out of date metadata and remove them.
```
Comment (rthille) [inline afkak/client.py]: Hmmm, I found this comment confusing at first. What you mean is a broker where the current cluster metadata has a `broker id -> (host, port)` mapping that doesn't match our KafkaBrokerClient's mapping, right?  Not sure if there's a better way to write that, but I was confused by the idea that a KBC "has" metadata...

File: afkak/client.py [hunk 3]
```diff
@@ -652,42 +660,61 @@ def _clean_close_dlist(result, close_dlist):
...
+            for every broker given if it doesn't yet exist.
+        :param bool remove:
+            Is this metadata for *all* brokers? If so, clients for brokers
+            which are no longer found in the metadata will be closed.
+        """
+        log.debug("%r: _update_brokers(%r, remove=%r)",
+                  self, brokers, remove)
+        brokers_by_id = {bm.node_id: bm for bm in brokers}
+        added_brokers = set(brokers_by_id) - set(self._brokers)
+        removed_brokers = set(self._brokers) - set(brokers_by_id)
```
Comment (rthille) [inline afkak/client.py]: Worth preserving the intermediate sets?  You're calculating both 2x.

File: afkak/client.py [hunk 4]
```diff
@@ -652,42 +660,61 @@ def _clean_close_dlist(result, close_dlist):
...
-        added_brokers = new_brokers - current_brokers
-        # removed
-        removed_brokers = current_brokers - new_brokers
+        to_close = []
 
-        # Create any new brokers based on the new metadata
-        for broker in added_brokers:
-            self._get_brokerclient(*broker)
+        self._brokers.update(brokers_by_id)
+        for node_id in removed_brokers:
```
Comment (rthille) [inline afkak/client.py]: Why isn't this protected by 'if remove'?  And why are you doing basically the same thing twice?

File: afkak/client.py (hunk 3 above)
Comment (twm) [reply afkak/client.py]: I may optimize once I write tests, but I'm preferring clarity for now. I doubt it really matters though.

File: afkak/client.py (hunk 2 above)
Comment (twm) [reply afkak/client.py]: Yeah all of this logic is super mushy as I haven't fully thought it through yet. That'll come when I write tests for it.

**Comment by twm** [comment]
Yes, I have a fix for that. I am working on the tests for this.

**Comment by twm** [comment]
@rthille No worries. I will find someone internally. While the changeset seems to work in practice, I still have some unit tests to fix up before then, though. This has ballooned.

File: afkak/test/testutil.py [hunk 5]
```diff
@@ -41,20 +55,48 @@ def stat(key, value):
...
 
-    timeout = timeout
-
-    def succeed():
-        d.callback(timeout)
-
-    d = Deferred()
-    clock.callLater(timeout, succeed)
-    return d
+    return task.deferLater(clock, timeout, lambda: None)
```
Comment (twm) [inline afkak/test/testutil.py]: This makes the deferred returned by async_delay cancellable, so it can't pollute the reactor.

**Comment by twm** [comment]
Thank you for the reviews, @rthille and @ryban!

**Pull request merged**